examples/agentbasedSEIR.jl

```diff
@@ -0,0 +1,180 @@
+
+# - states: a collection of distinct states an agent can occupy
+# - agents: a collection of states `aᵢ = agents[i] ∈ states` indicating that agent `i` is in state `aᵢ`
+# - transitions: the functions `f: states -> states`
+# """
+
+mutable struct StateModel <: AgentModel
+    states
+    agents
+    transitions
+    loads
+end
+
+# +
+# Determine number of agents in a given state
+function count(sm::StateModel, state)
+    return length(filter(a->(a==state), sm.agents))
+end
+
+# counts the number of agents
+function count(sm::StateModel)
+    return length(sm.agents)
+end
+# -
+
+# stateload computes the fraction of agents in each state
+# it is used by tick! to update the statemodel for computing the probability of infection.
+function stateload(sm::StateModel, state::Symbol)
+    return (count(sm, state)+1)/(count(sm)+1)
+end
+
+#     tick!(sm::StateModel)
+#
+# performs the operations that need to happen once per time step. Things like caching values that need to be computed at the beginning of the timestep like the distribution of the agents' states.
+
+function tick!(sm::StateModel)
+    sm.loads = map(s->stateload(sm, s), sm.states)
+    return sm.loads
+end
+
+#     step!(sm::StateModel, n)
+#
+# advance the simulation by `n` ticks of time.
+# This is an in-place operation that modifies the current state of the simulation.
+#
+
+function step!(sm::StateModel, n)
+    for s in 1:n
+      tick!(sm)
+      for (i, a) in enumerate(sm.agents)
+          sm.agents[i] = sm.transitions(sm, i, a)
+      end
+        describe(sm)
+    end
+    return sm
+end
+
+# """    describe(sm::StateModel)
+
+# summarize the state of the simulation for presentation or analysis.
+# """
+
+function describe(sm::StateModel)
+    counts = zeros(Int, size(sm.states))
+    d = Dict{eltype(sm.states), Int}()
+    for (i,s) in enumerate(sm.states)
+        d[s] = i
+    end
+    for a in sm.agents
+        i = d[a]
+        counts[i] += 1
+    end
+    return collect(map(x->Pair(x...), zip(sm.states, counts)))
+end
+
+# ## Run the model
+#
+# This script defines a basic agent based model of disease spread called SEIRS. Each agent is in one of 4 states
+#
+# 1. $S$ Susceptible
+# 2. $E$ Exposed
+# 3. $I$ Infected
+# 4. $R$ Recovered
+#
+# The agents go from `S->E`, `E->I`, `I-R`, and `R->S` based on random numbers. The probability of S-> is dependent on the fraction of agents in state :E. The probability of recovering is a constant ρ, the probability of exposure is denoted by variable β, and the disease confers some temporary immunity with probability μ.
+#
+# This script has an entrypoint to call it so that you can include this file and run as many simulations as you want. The intended use case is to repeatedly call `main` and accumulate the return values into an array for later analysis.
+
+ρ = 0.5 + randn(Float64)/4 # chance of recovery
+μ = 0.5 # chance of immunity
+β = 0.5 # chance of exposure
+function transition(sm::StateModel, i::Int, s::Symbol)
```

Comment:
So this has one function with a bunch of switching logic. The other version has one function per state and uses a dictionary to implement this switching logic. Which is better?

Reply:
Perhaps in terms of cleanliness the dictionary implementation would be better. But may be harder to demonstrate in an automated flow in the demo. But I can do both, if you would like this in agentbased.jl?

Reply:
A suggestion: function per state is something we can handle immediately; we are still working on structured types. We're just finishing work on arrays, dictionaries (hash) are next.

Reply:
Ok that is great feedback @cl4ton. We should go with the function per state approach.